### DISCUSSION TITLE
Field visibility dependent on a field in table related through relationship

### DISCUSSION THREAD

**Discussion by phaethon** [Budibase/budibase#8817] [Help]
I have a following use case. I have a form which contains a relationship picker. In the table linked through this relationship there is a boolean column (let's call it show_field). Depending on this value another field needs to be shown or not. If value of relationship is changed then the field is correspondingly hidden or shown.
Data source is PostgreSQL.
First, I tried to create a Data provider and Repeater for relationship table and define a filter for data provider with id equals value of relationship field. This did not work.
Then I tried to compare Relationship picker field value to _id field. I got error message that table does not contain _id (even though I could choose it through drop down).
Quick check of idea of creating a formula field for the table - it does not work as the form is not saved at that moment and formula is not calculated.
Then I started to investigate what kind of value is returned by Relationship. Apparently, when the form is opened it returns an object with field _id containing the value of _id, on subsequent changes of relationship picker its value becomes just a value of _id.

Long story short - I found that comparing id field of the table with result of this function works.
```
const r = $("Time Form.Fields.activities");
let internal_id;

if (Object.hasOwn(r[0], '_id')) {
	internal_id = r[0]['_id'];
} else {
	internal_id = r[0];
}

return parseInt(internal_id.substring(3, internal_id.length - 3));
```

But it feels like a hack. Is there better and more straight forward way to achieve the result?

**Comment by melohagan** [comment]
Hey @phaethon 

Had you looked at this docs page at all? https://docs.budibase.com/docs/filter-by-relationships

**Reply by phaethon** [reply]
Yes, @melohagan I did read the page, but it does not work as described (or I don''t understand it).

Let's call call table I use for the form itself Main, and table I use for values of the Relationship picker as Attribute. Main and Attribute are related as many -> one, and there is a relationship defined. Attribute table contains a field show_attribute2. When an attribute is selected, which has corresponding show_attribute2 set to True, then Form needs to show field attribute2 (or no if it is False).
I have a data provider set to Main table, Repeater with provider set to Main data provider, and form with schema set to Main. 
Inside the form there is relationship picker and field attribute2.

I did try following:
1. Having Data provider inside the form with data set to Repeater.attribute. Add new repeater inside it (AttributeRepeater). Have field attribute2 inside this repeater. And add a condition to attribute2: 
Show component IF {{ AttributeRepeater.attribute.show_attribute2 }} EQUALS Boolean True
Result: field is never shown
2. Having Data provider with data set to Form.attribute. Have repeater inside it (AttributeRepeater). And attribute2 inside this repeater.
And add a condition to attribute2: 
Show component IF {{ AttributeRepeater.attribute.show_attribute2 }} EQUALS Boolean True
Result: Circle rotating instead of field being shown (both as part of design screen and when I use Preview).

What did I misunderstand from the documentation article?

**Reply by melohagan** [reply]
> Then I tried to compare Relationship picker field value to _id field. I got error message that table does not contain _id (even though I could choose it through drop down).

This is a known bug, sorry! https://github.com/Budibase/budibase/issues/8334
The workaround is to use the actual primary key field instead, e.g. *product_id*

---

What you tried in the first instance seems correct to me (assuming the Data provider is filtered to output only one row). 
Is *show_attribute2* definitely defined as a boolean in your schema?

![Screenshot 2022-11-28 at 09 02 36](https://user-images.githubusercontent.com/101575380/204236632-ff5c2e1c-2239-44aa-b393-08efd4537239.png)

**Reply by phaethon** [reply]
Problem with id/_id fields is that Relationship picker value is _id, but table contains only id (taking into account #8334). So, the workaround from #8334 is not applicable 1:1 here. To compare these 2 I ended up using the javascript code included in the start of this discussion. 

Yes, show_attribute2 is defined as boolean.

To debug further the issue with "first instance" I added a table to the form to see which rows are returned. Contents of the table are correct when the form is loaded, but the value does not change when I change the picker value (which is logical as the form is not yet saved). Previously I wrote "field is never shown", it is not correct - it works, but only when loading form and ignores subsequent changes.
